src/Router.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+namespace Zap\Routing;
+
+use Zap\Routing\Exceptions\InvalidRouteException;
+
+/**
+ * Class Router
+ * @package Zap\Routing
+ * @author Gabor Zelei
+ */
+class Router
+{
+    /**
+     * @var Router
+     */
+    private static $instance;
+
+    /**
+     * @var Matcher
+     */
+    private $matcher;
+
+    /**
+     * @var RouteCollection
+     */
+    private $routes;
+
+    /**
+     * Router singleton constructor.
+     */
+    private function __construct()
+    {
+        $this->matcher = new Matcher($this);
+        $this->routes = new RouteCollection();
+    }
+
+    /**
+     * Neuter clone method
+     */
+    private function __clone()
+    {
+    }
+
+    /**
+     * Singleton instance accessor
+     * @return Router
+     */
+    public static function get() : Router
+    {
+        if (is_null(static::$instance)) {
+            static::$instance = new static();
+        }
+
+        return static::$instance;
+    }
+
+    /**
+     * Adds a route to the collection this router knows about
+     * @param Route $route
+     * @return Router
+     */
+    public function addRoute(Route $route) : Router
+    {
+        $this->routes->add($route);
+        return $this;
+    }
+
+    /**
+     * Attempts to find a route based on current uri and request method
+     * @param string $uri
+     * @param string $method
+     * @return mixed|null
+     */
+    public function fetch(string $uri, string $method)
```

Comment:
This could get as argument a `Response`, which contains both the uri and the method.